Adding a helper function to check if should be verified.

eloycoto opened the pull request:
When handshake happens, we also need to check the CN. Something that
it's not done because u->conf->verify maybe is 0. With this change, a
new function is exported to handle this case on nginx-upstream.

Lines that need to be checked:
https://github.com/nginx/nginx/blob/02bd43d05b6f7803597d8453d9848b767dc4a323/src/http/ngx_http_upstream.c#L1786-L1789

Changes that need to be done on Openresty:
```

static void
ngx_http_upstream_ssl_handshake(ngx_http_request_t *r, ngx_http_upstream_t *u,
    ngx_connection_t *c)
{
    long  rc;

    if (c->ssl->handshaked) {

        if (u->conf->ssl_verify || ngx_http_upstream_should_verified(r) == NGX_OK) {
```

Signed-off-by: Eloy Coto <eloy.coto@acalustra.com>

unleashed reviewed (commented):
So this requires us to add a function here that isn't used by this module but called from the nginx sources directly, and we will need to patch nginx anyway - why don't we just patch nginx adding this code instead?

eloycoto commented:
The thing is, I want to keep our nginx patches as minimal as possible, how the data is shared (ctx) in the module is much easier, and it's better to provide support in the future. 

Regards

unleashed commented:
@eloycoto I don't follow what is easier and what would be better about support. Can you elaborate?

As I see it this is adding not only a patch to the sources, but also a direct dependency to the module sources, and an extraneous function to the module that apparently is not used locally (so it would be easy for anyone to just think it is not needed and proceed to remove it), whereas you can just have a patch to the sources without depending on the module, at least for the C sources shown here. A patch and a dependency on a function that the module does not need looks strictly worse to me than a function that can be self contained in the patch.

Is there any issue in incorporating the body of this function to the nginx sources?

That said, the need to patch this other upstream module looks suspicious to me - I lack the context to understand why what it does is not enough, and I think that would help me understand better the needs, so I'd like this PR to point out why it is not enough and whether an issue/PR has been opened *on the upstream repos* (nginx/openresty) so the module is fixed in the future, assuming it makes sense to do that -- or maybe it doesn't or the changes won't be accepted, but still good to document it here and in the code/patch so I and others can pick up the context and understand the reasons we do this.

eloycoto commented:
Ok, after addressing some of your comments on private. I added a way to build openresty and the module, so things are now much much easier to build, follow. 

Regards

unleashed commented:
Thanks @eloycoto, I'll check this out ASAP.

unleashed reviewed (approved):
Ok, this is a nice addition.
